Quimb Tensor Network Simulators

mpharrigan opened the pull request:
[Quimb](https://quimb.readthedocs.io/en/latest/index.html) is a nice tensor network package. This set of utilities (in `contrib` because there are features missing) allows you to turn cirq Circuits into quimb tensor networks with some benefits.

![image](https://user-images.githubusercontent.com/4967059/87093421-05b1b480-c1f2-11ea-8804-2f78b7cb3620.png)

![image](https://user-images.githubusercontent.com/4967059/87093437-0ba79580-c1f2-11ea-87c0-d7e6617df789.png)


The particular use cases I've explored explicitly is

1. For shallow, low degree circuits you can compute the value of low weight observables on a very high number of qubits (example provided in "Contract a Grid Circuit" notebook)
2. For low-qubit number, deep, noisy circuits it is faster than the cirq density matrix simulator

![image](https://user-images.githubusercontent.com/4967059/87093490-22e68300-c1f2-11ea-8953-f8f5f59abac3.png)


## Features

 - Turn circuits into tensor networks representing state vector or unitary
 - Drop-in replacements for `cirq.final_state_vector` and `cirq.unitary` that use tensor contraction under the hood
 - Utilities for contracting circuits of the form <0 | U^dag A U |0> for PauliString `A`. 
 - Some related functionality for building circuits on a grid and simplifying some of the operations with low weight observables
 - Circuit-like positioning of tensors in tensor drawings for density matrix networks

## Limitations

 - Code duplication between state vector and density transforms
 - Initial state must be |0>
 - No circuit-like positioning of tensors in state-vector networks
 - No expectation values with density matrices
 - Doesn't follow the simulator and sampler APIs

Strilanc commented on cirq/contrib/quimb/density_matrix.py:
```diff
@@ -0,0 +1,195 @@
+    Args:
+        circuit: The circuit containing operations that support the
+            cirq.unitary() or cirq.channel() protocols.
+        qubits: The qubits in the circuit.
```
Can this be an optional argument, defaulting to `sorted(circuit.all_qubits())`?

mpharrigan replied:
done here and below with test

Strilanc commented on cirq/contrib/quimb/density_matrix.py:
```diff
@@ -0,0 +1,195 @@
+    return tensors, qubit_frontier, positions
+
+
+def tensor_density_matrix(circuit, qubits):
```
Type annotations

mpharrigan replied:
added

Strilanc commented on cirq/contrib/quimb/density_matrix.py:
```diff
@@ -0,0 +1,195 @@
+        (mi * x_scale, y_scale * qy + yb_offset)
+
+
+def circuit_to_density_matrix_tensors(circuit: cirq.Circuit,
```
Type annotation for result

mpharrigan replied:
added

Strilanc commented on cirq/contrib/quimb/state_vector.py:
```diff
@@ -0,0 +1,144 @@
+import cirq
+
+
+def circuit_to_tensors(circuit: cirq.Circuit,
```
Result type annotation

Strilanc commented on cirq/contrib/quimb/state_vector.py:
```diff
@@ -0,0 +1,144 @@
+    return tensors, qubit_frontier, positions
+
+
+def tensor_state_vector(circuit, qubits):
```
type annotations

mpharrigan replied:
added

Strilanc commented on cirq/contrib/quimb/state_vector.py:
```diff
@@ -0,0 +1,144 @@
+    return tn.to_dense(f_inds)
+
+
+def tensor_unitary(circuit, qubits):
```
type annotations

mpharrigan replied:
added

Strilanc commented on cirq/contrib/quimb/state_vector.py:
```diff
@@ -0,0 +1,144 @@
+    return tn.to_dense(f_inds, i_inds)
+
+
+def circuit_for_expectation_value(circuit: cirq.Circuit,
```
result type annotation

balopat commented on cirq/contrib/quimb/Contract-a-Grid-Circuit.ipynb:
```diff
@@ -0,0 +1,405 @@
+    "%matplotlib inline\n",
+    "from matplotlib import pyplot as plt\n",
+    "\n",
+    "import seaborn as sns\n",
```
seaborn is missing from the contrib requirements - maybe worth adding it or at least mentioning it?

balopat commented on cirq/contrib/quimb/density_matrix.py:
```diff
@@ -0,0 +1,195 @@
+def _add_to_positions(positions: Dict[Tuple[str, str], Tuple[float, float]],
+                      mi: int,
+                      qubits: Union[cirq.LineQubit, Tuple[cirq.LineQubit]], *,
+                      n_qubits: int, x_scale, y_scale, x_nudge, yb_offset):
```
isn't n_qubits == len(qubits)?

mpharrigan replied:
see other comment; changed name; updated docstring

balopat commented on cirq/contrib/quimb/density_matrix.py:
```diff
@@ -0,0 +1,195 @@
+
+@lru_cache()
+def _qpos_y(qubits: Union[cirq.LineQubit, Tuple[cirq.LineQubit]],
+            n_qubits: int):
```
isn't n_qubits == len(qubits)?

mpharrigan replied:
No. Changing `n_qubits` to `tot_n_qubits`, the total number of qubits (needed to turn qubit index into a y position (which are backwards)). `qubits` is usually one or two qubits, corresponding to a tensor, corresponding to a gate operation

mpharrigan replied:
and adding a docstring

mpharrigan commented on cirq/contrib/quimb/state_vector.py:
```diff
@@ -0,0 +1,144 @@
+        qubit_frontier: A mapping from qubit to time index at the end of
+            the circuit. This can be used to deduce the names of the free
+            tensor indices.
+        positions: Currently None. May be changed in the future to return
```
I'm going to push back for the following three reasons

1. currently lets you switch between `circuit_to_tensors` and `circuit_to_density_matrix_tensors` with no other code changes
2. quimb utilities consuming this result handle `None` gracefully
3. I'm pretty sure this is going to get implemented, and changing return values is hard to do 
4. This function is a little lower-level than `tensor_unitary` and `tensor_state_vector`, and as such the user can probably handle dealing with the more "under-the-hood" tensors and qubit frontier (and positions)

balopat replied:
Sounds good, you convinced me

mpharrigan commented on cirq/contrib/quimb/Cirq-to-Tensor-Networks.ipynb:
```diff
@@ -0,0 +1,364 @@
+   "metadata": {},
+   "source": [
+    "### Circuit to Tensors\n",
+    "The circuit defines a tensor network representation. By default, the initial state is the `|0...0>` state (represented by the \"zero qubit\" operations labeled \"Q0\" in the legend. \"Q1\" are single qubit operations and \"Q2\" are two qubit operations. The open legs are the indexes into the state vector and are of the form \"i{m}_q{n}\" where `m` is the time index (given by the returned `qu ...
```
done! (locally)

mpharrigan commented on cirq/contrib/quimb/density_matrix.py:
```diff
@@ -0,0 +1,195 @@
+    """Given a circuit with mixtures or channels, construct a tensor network
+    representation of the density matrix.
+
+    This assumes you start in the |0..0><0..| state. Indices are named
```
fixed (locally!). Thanks for the spot

mpharrigan commented:
I addressed all the comments and the build is green. I see you both gave approval, but I'll give a couple hours for any follow up reviews before I merge

balopat commented:
From my end it looks good!
